rust_build_utils/msvc.py

```diff
@@ -0,0 +1,149 @@
+import subprocess
+from contextlib import contextmanager
+from pathlib import Path
+import re
+import os
+from typing import Optional
+
+
+@contextmanager
+def activate_msvc(

```

Comment:
In general I'm not sure this should be a reponsibility of a build script to prepare an environment like this. It is machine dependent and I also feel like one should be able to pick version they want to use instead of relying on some arbitrary ordering. But I see a value which is just an ease of use for someone who is not used to working with msvc. Still I'm not sure if this is enough of an argument for me

Reply:
> In general I'm not sure this should be a reponsibility of a build script to prepare an environment like this. It is machine dependent and I also feel like one should be able to pick version they want to use instead of relying on some arbitrary ordering. But I see a value which is just an ease of use for someone who is not used to working with msvc. Still I'm not sure if this is enough of an argument for me

Ordering can be changed based on `*_preference` parameters. As far as responsibilities go I'm very conflicted myself. Ideally any user of build utils would activate the msvc environment before calling the build functions. But the environment is required only for specific Rust targets, meaning that the API of build scripts would read "you need to call `activate_msvc` before calling `cargo_build` if `config.target` is "-msvc" - it would introduce a required function call for some parameters, meaning those are not really independent parameters. Secondly, our build scripts do make a lot of assumptions about where they will run - firstly that Rust is installed and added to the path, secondly about the version of Rust used (and NDK version in the case of android scripts). Their purpose is not be a general purpose scripts (for that the raw tools are the only complete fit), but rather abstract away *our* infrastructure, which does include specific machines.

Reply:
I think this script for detecting and sourcing VS vars is great. Its always a pain on Windows to find the correct path to these `vars` scripts. I guess it would be great if it was possible to somehow override this path on the runner instance by adding some env, in case the runner has multiple VS installations and the default ordering in the script is not sufficient. But I think this is an edge case, and picking the latest installed version is a great default.

Reply:
> Ordering can be changed based on *_preference parameters
It can be changed but it's not exposed in public API of rust_utils. So if I just run `cargo_build` from my project, I don't have a way to influence this choice.

> API of build scripts would read "you need to call activate_msvc before calling cargo_build
I don't think this is an API thing. I do not propose to ask users to call activate_msvc  themselves in the script. I just say that maybe it should not be a part of this script at all. Like you said in all other instances we assume environment is correctly configured. e.g. rust is always in a path, if building for android android ndk is in the path. Why should we treat msvc differently instead of assuming that cl.exe, link.exe etc are in the path before someone called the script. I would simply check if they are and if not I would inform the user that if they want to use msvc toochain they need to have msvc tools in path.
Maybe it could be also done this way: if the tools are not set then use activate_msvc. If the user have msvc tools in the path then skip it. This way activate_msvc can be simplified. No need to allow user to influence it's choices. For 99% of them the default is ok, and the rest can just set the environment themselves.

Reply:
> that doesn't stop you from using msvc module individually

I think in this form it does stop me, because even if I set the env myself, when I run `cargo_build` with `-msvc` target, the `activate_msvc` function will be called anyway, so if there is some issue with this function I don't have an easy way to skip this step. That's why I would propose to add some check that if env is set already do not execute this function, or I would remove this step entirely (That's actually my opinion. That the only thing this PR should do is to pass target argument to the cargo invocation)

> the whole if target.ends_with("-msvc"): activate_msvc() will go away

I think I don't get this part. What is stopping you currently from removing this part? What would the API that would allow this?

Reply:
I just realized that automatically activating the environment is especially nice for local use. Building libtelio on Windows machine fails without explicitly sourcing or using the special VS pre-sourced terminal.

Reply:
Yeah, it's convenient but it might be annoying if it fails. But now there is `is_msvc_active()` function which allows users to initialize msvc toolchain on their own before hand and the script will not fail, so I'm OK with it now.